#696, New captcha window

K1rakishou opened the pull request:
Closes #696 
For some reason it takes me like 20 tries to solve the captcha maybe because I tested it a lot today and it just doesn't like me now. Or maybe I forgot to add something (I did add the SID, HSID, SSID and NID cookies though). It should automatically fallback to the webview captcha if something couldn't be parsed (or you can use the button).

![345345](https://user-images.githubusercontent.com/10332458/56867678-b7e7cc80-69f0-11e9-8660-30faf7ffe738.png)

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/ui/captcha/v2/CaptchaNoJsHtmlParser.java:
```diff
@@ -0,0 +1,398 @@
+            }
+
+            return resultImages;
+        } finally {
```
Should recycle bitmaps (in the catch clause) that were already added to the list when an exception occurs in the process of decoding the images. Right now they will leak.

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/ui/captcha/v2/CaptchaNoJsLayoutV2.java:
```diff
@@ -0,0 +1,274 @@
+    public void onCaptchaInfoParsed(CaptchaInfo captchaInfo) {
+        // called on a background thread
+
+        AndroidUtils.runOnUiThread(() -> {
```
Should probably disable the Verify button until we have the CaptchaInfo. Do not forget to reset the button's disabled state in initialize/onError/onVerificationDone/onDestroy.

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/ui/captcha/v2/CaptchaNoJsPresenterV2.java:
```diff
@@ -0,0 +1,347 @@
+                public void onFailure(Call call, IOException e) {
+                    if (callbacks != null) {
+                        try {
+                            callbacks.onCaptchaInfoParseError(e);
```
Set prevCaptchaInfo to null here

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/ui/captcha/v2/CaptchaNoJsPresenterV2.java:
```diff
@@ -0,0 +1,347 @@
+                public void onFailure(Call call, IOException e) {
+                    if (callbacks != null) {
+                        try {
+                            callbacks.onCaptchaInfoParseError(e);
```
Set prevCaptchaInfo to null here

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/ui/captcha/v2/CaptchaNoJsPresenterV2.java:
```diff
@@ -0,0 +1,347 @@
+                    callbacks.onCaptchaInfoParsed(captchaInfo);
+                }
+
+                return captchaInfo;
```
Probably should return null here to reset prevCaptchaInfo when callbacks are null. Otherwise we are kinda stuck in a state where there is no captcha window (But I guess reset should fix it?).

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/ui/layout/ReplyLayout.java:
```diff
@@ -269,7 +272,25 @@ public void initializeAuthentication(Site site, SiteAuthentication authenticatio
                 }
                 case CAPTCHA2_NOJS:
-                    authenticationLayout = new CaptchaNojsLayout(getContext());
+                    if (!useV2NoJsCaptcha) {
```
Probably should invert the condition here

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/ui/layout/ReplyLayout.java:
```diff
@@ -269,7 +272,25 @@ public void initializeAuthentication(Site site, SiteAuthentication authenticatio
+
+                    ImageView resetButton = captchaContainer.findViewById(R.id.reset);
+                    if (resetButton != null) {
+                        if (!useV2NoJsCaptcha) {
```
Probably should invert the condition here

K1rakishou commented on Clover/app/src/main/res/layout/layout_captcha_challenge_image.xml:
```diff
@@ -0,0 +1,29 @@
+        android:layout_marginTop="8dp"
+        android:visibility="gone"
+        android:src="@drawable/ic_blue_checkmark_24dp"
+        app:layout_constraintStart_toStartOf="@+id/captcha_challenge_image"
```
Add #FFF tint here

K1rakishou replied:
Implemented it the other way

Adamantcheese commented:
@K1rakishou I submitted a PR to fix the checkmark not being centered, and having to use an extra background view to make it white due to transparency.

K1rakishou commented:
@Adamantcheese 
>checkmark not being centered

But that's how it is implemented in the google's recaptcha window. You think it should be centered?
![234234](https://user-images.githubusercontent.com/10332458/57431237-d8a6f200-723a-11e9-89a2-189513f2b665.png)


>to make it white due to transparency

And this one already fixed here - https://github.com/Floens/Clover/pull/697/commits/2820d23ff787e7d97cb69c0536ea5e2d08c8e3fb#diff-460f4eca5fc92d000ad076afc5756ccaR31

Adamantcheese commented:
The white thing I changed so it's part of the drawable and doesn't use an underlying view to make it white, because that's just silly. And for me that check mark was appearing... elsewhere? Here's a photo of the Pixel 2 emulator from when I tested it. It looks like you've crafted it specifically for the second style of captcha and not the first one, and I pretty much only ever get the 3x3 kind and not the 4x4 kind.
![](https://i.imgur.com/5DNsPmc.png)

Adamantcheese commented:
And yeah, I had to enter the Captcha like 5-6 times.

K1rakishou commented:
@Adamantcheese 
>The white thing I changed so it's part of the drawable and doesn't use an underlying view to make it white, because that's just silly.

Oh yeah, this makes sence.

>And for me that check mark was appearing... elsewhere? Here's a photo of the Pixel 2 emulator from when I tested it. It looks like you've crafted it specifically for the second style of captcha and not the first one, and I pretty much only ever get the 3x3 kind and not the 4x4 kind.

Well that's strange because I only get the 3x3 captchas too and the checkmark is where it is supposed to be (upper left corner).

K1rakishou commented:
Alright then I will merge it later today. Still we have to do something with the amount of captcha attempts it takes to make a post.

K1rakishou commented:
@Adamantcheese 
Over the last week I tried it on the real device too and most of the time it gives you the same amount of challenges. But sometimes it just accepts the first one.

Adamantcheese commented:
Weird. I don't know anything about how Captcha works though, so I won't be much help.

K1rakishou commented:
Well, I guess for now we can just make it opt-out.
The other idea I have is to get the real google cookies upon the first app run, then store them somewhere and use them instead of the hardcoded ones.

Adamantcheese commented:
From a little research it looks like the cookies might be per-user, which would make the hardcoded ones a problem anyways because anyone using the app would be using the same cookie and they might be around the world, which looks sketchy as hell.

K1rakishou commented:
Yeah, probably. But on the other hand dashchan uses hardcoded cookies and works just fine.

Adamantcheese commented:
Weird. No idea then.

K1rakishou commented:
Alright. I added two new settings to enable/disable the new captcha window and to use the real or the default google cookie. In case of using the real cookie a GET request to the google.com will be made to retrieve the NID cookie (apparently it is the most important among all of them) then this cookie will be used for recaptcha requests until app restart. Also I added a button to manually refresh the cookie (sometimes it helps when you are stuck in the challenge hell). It does look like it helps a little bit when using the real google cookie.

K1rakishou commented on Clover/app/src/main/res/values/strings.xml:
```diff
@@ -577,4 +577,19 @@ Don't have a 4chan Pass?<br>
+    <string name="settings_captcha_group" tools:ignore="MissingTranslation">Captcha</string>
+    <string name="settings_use_new_captcha_window" tools:ignore="MissingTranslation">Use new captcha window for no-js captcha</string>
+    <string name="settings_use_real_google_cookies" tools:ignore="MissingTranslation">Use real google cookies instead of hardcoded ones</string>
+    <string name="settings_use_real_google_cookies_description" tools:ignore="MissingTranslation">When the real google cookies a GET request to the google.com will be executed to get the google cookies (NID) which will be used for captcha authentication. Why would you need them? Because the hardcoded ones sometimes will make you re-enter the captcha dozens of times. Those cookies will be updated  ...
```
Fix message

K1rakishou commented:
Instead of refreshing google cookie (when this option is enabled) on every app start I decided to do it once in three months (but there is still an ability to refresh it manually). I think it is a better idea to not refresh them way too often.
